### PR TITLE
Redesign button component #35

### PR DESCRIPTION
**Pull request by Atzac**
## Description

Refactoring of the primary button, initially created for testing, and added 3 more variants:

- Secondary
- Ghost
- dashed

### CODE REVIEW

File: src/components/button/button.test.tsx [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
+  test('execute the onClick event', () => {
+    const handleClick = jest.fn();
+
+    render(<Button type="primary" label="Button" onClick={handleClick} />);
+
+    fireEvent.click(screen.getByRole('button', { name: 'Button' }));
+
+    // expect mock handler to be called
+    expect(handleClick).toBeCalledTimes(1);
+  });
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:22]: Create new tests cases to check styles

File: src/components/button/button.tsx [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
-export interface ButtonProps extends InputHTMLAttributes<HTMLButtonElement> {
-  label: string;
-}
+type StitchesButtonProps = React.ComponentProps<typeof StitchesButton>;
 
-const Button = ({ label }: ButtonProps) => {
-  return <StitcheButton>{label}</StitcheButton>;
+export type ButtonProps = {
+  label?: string;
+  type: string;
```
Comment (iurynogueira) [inline src/components/button/button.tsx:8]:
```suggestion
  type: string;
```

Primary can be default

File: src/components/button/styles.ts [hunk 3]
```diff
@@ -2,22 +2,105 @@ import stitches from '../../stitches.config';
...
+    color: fontColor,
+  };
+
+  borderColor && (styles.border = `1px solid ${borderColor}`);
+
+  return styles;
+}
+
+// Componente Button
+export const StitchesButton = styled('button', {
```
Comment (iurynogueira) [inline src/components/button/styles.ts:37]: Change style name

File: src/components/button/styles.ts [hunk 4]
```diff
@@ -2,22 +2,105 @@ import stitches from '../../stitches.config';
...
+  minWidth: '32px',
+  height: '32px',
+  gap: '8px',
+
   variants: {
-    size: {
-      sm: {
-        width: '75px',
+    type: {
+      primary: {
```
Comment (iurynogueira) [inline src/components/button/styles.ts:51]: danger is missing

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: src/components/button/button.test.tsx (hunk 1 above)
Comment (Atzac) [reply src/components/button/button.test.tsx:50]: adjusted

File: package.json [hunk 5]
```diff
@@ -100,6 +100,7 @@
     }
   },
   "dependencies": {
-    "@stitches/react": "^1.2.8"
+    "@stitches/react": "^1.2.8",
+    "@testing-library/jest-dom": "^5.16.5"
```
Comment (iurynogueira) [inline package.json:104]: 👍

File: src/components/button/button.test.tsx [hunk 6]
```diff
@@ -1,12 +1,51 @@
 import React from 'react';
-import { render, screen } from '@testing-library/react';
+import { render, screen, fireEvent } from '@testing-library/react';
+import '@testing-library/jest-dom';
 
 import Button from './button';
+import { ButtonProps } from './button';
+
+function renderAndGetButton(props: ButtonProps): HTMLElement {
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:8]:
```suggestion
function sut(props: ButtonProps): HTMLElement {
```

https://github.com/Brisanet/typescript/blob/main/docs/Testes%20Unitarios.md#system-under-test-sut

keep this pattern

Comment (iurynogueira) [inline src/components/button/button.test.tsx:8]:
```suggestion
function renderAndGetButton(props: ButtonProps = {...defaultProps}): HTMLElement {
```

File: src/components/button/button.test.tsx [hunk 7]
```diff
@@ -1,12 +1,51 @@
...
+  render(<Button label={props.label || 'Button'} {...props} />);
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:9]:
```suggestion
  render(<Button {...props} />);
```

You can do this when change above

File: src/components/button/button.test.tsx [hunk 8]
```diff
@@ -1,12 +1,51 @@
...
+  return screen.getByRole('button', { name: props.label });
+}
 
 describe('Button', () => {
   test('renders the Button component', () => {
     const label = 'Hello world!';
-    render(<Button label={label} />);
+    render(<Button type="primary" label={label} />);
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:16]:
```suggestion
    render(<Button label={label} />);
```

primary is default

File: src/components/button/button.test.tsx [hunk 9]
```diff
@@ -1,12 +1,51 @@
...
     expect(screen.queryAllByText(label)).toHaveLength(1);
   });
+
+  test('execute the onClick event', () => {
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:20]:
```suggestion
  test('execute the onClick event', () => {
```

Prefer start spec with `should`

Comment (iurynogueira) [inline src/components/button/button.test.tsx:20]:
```suggestion
  test('should execute event when the button is clicked', () => {
```

File: src/components/button/button.test.tsx [hunk 10]
```diff
@@ -1,12 +1,51 @@
...
+    expect(button).toHaveAttribute('disabled');
+  });
+
+  const buttonTypes: Array<ButtonProps['type']> = [
+    'primary',
+    'secondary',
+    'ghost',
+    'dashed',
+  ];
+  test.each(buttonTypes)('should render all types', async (type) => {
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:45]:
```suggestion
  test.each(buttonTypes)('should render button with %s class', async (type) => {
```

File: src/components/button/button.test.tsx [hunk 11]
```diff
@@ -1,12 +1,51 @@
...
+    expect(renderAndGetButton({ type })).toHaveAttribute(
+      'class',
+      expect.stringContaining(`type-${type}`)
+    );
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:49]: ToHaveClass

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (Atzac) [reply package.json:104]: @iurynogueira  all ok here?

File: src/components/button/button.tsx [hunk 12]
```diff
@@ -1,12 +1,33 @@
...
-  return <StitcheButton>{label}</StitcheButton>;
+export type ButtonProps = {
+  label?: string;
+  type?: string;
+  disabled?: boolean;
+  onClick?: () => void;
+  danger?: boolean;
+} & StitchesButtonProps;
+
+const Button = ({
```
Comment (Deeved) [inline src/components/button/button.tsx:14]:
```suggestion
const IonButton = ({
```

File: src/components/button/button.tsx [hunk 13]
```diff
@@ -1,12 +1,33 @@
...
+      onClick={onClick}
+      disabled={disabled}
+      className={`${danger && 'danger'}`}
+    >
+      {label}
+    </ButtonStyles>
+  );
 };
 
 export default Button;
```
Comment (Deeved) [inline src/components/button/button.tsx:33]:
```suggestion
export default IonButton;
```

**Review by Deeved — APPROVED** [review]

File: src/components/button/button.tsx [hunk 14]
```diff
@@ -1,12 +1,33 @@
...
 
-export interface ButtonProps extends InputHTMLAttributes<HTMLButtonElement> {
-  label: string;
-}
+type StitchesButtonProps = React.ComponentProps<typeof ButtonStyles>;
 
-const Button = ({ label }: ButtonProps) => {
-  return <StitcheButton>{label}</StitcheButton>;
+export type ButtonProps = {
+  label?: string;
```
Comment (iurynogueira) [inline src/components/button/button.tsx:7]:
```suggestion
  label: string;
```
label is required.

File: src/components/button/styles.ts [hunk 15]
```diff
@@ -2,22 +2,160 @@ import stitches from '../../stitches.config';
...
+      secondary: {
+        ...generateColorStyle('$neutral1', '$primaryColor', '$neutral4'),
+
+        '&:hover': generateColorStyle('$primary1', '$primary5', '$primary4'),
+        '&:active': generateColorStyle('$primary2', '$primary5', '$primary7'),
+        '&:disabled': {
+          ...generateColorStyle(
+            '$neutral2 !important',
+            '$neutral5 !important',
+            '$neutral5 !important'
```
Comment (iurynogueira) [inline src/components/button/styles.ts:85]: check use disabled as a variant

File: src/components/button/button.tsx [hunk 16]
```diff
@@ -1,12 +1,33 @@
...
+  disabled?: boolean;
+  onClick?: () => void;
+  danger?: boolean;
+} & StitchesButtonProps;
+
+const IonButton = ({
+  type = 'primary',
+  disabled,
+  onClick,
+  danger,
```
Comment (iurynogueira) [inline src/components/button/button.tsx:18]:
```suggestion
  disabled = false,
  onClick,
  danger = false,
```

File: src/stories/button.stories.tsx [hunk 17]
```diff
@@ -1,20 +1,45 @@
...
+export const ButtonPrimary = Template.bind({});
+export const ButtonSecondary = Template.bind({});
+export const ButtonGhost = Template.bind({});
+export const ButtonDashed = Template.bind({});
+
+ButtonPrimary.args = {
+  label: 'Primary',
+  type: 'primary',
+  danger: false,
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:23]:
```suggestion
```

File: src/stories/button.stories.tsx [hunk 18]
```diff
@@ -1,20 +1,45 @@
...
+};
+
+ButtonSecondary.args = {
+  label: 'Secondary',
+  type: 'secondary',
+  danger: false,
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:30]:
```suggestion
```

File: src/stories/button.stories.tsx [hunk 19]
```diff
@@ -1,20 +1,45 @@
...
+};
+
+ButtonGhost.args = {
+  label: 'Ghost',
+  type: 'ghost',
+  danger: false,
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:37]:
```suggestion
```

File: src/stories/button.stories.tsx [hunk 20]
```diff
@@ -1,20 +1,45 @@
...
+};
 
-ButtonDefault.args = {
-  label: 'Button',
+ButtonDashed.args = {
+  label: 'Dashed',
+  type: 'dashed',
+  danger: false,
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:44]:
```suggestion
```

File: src/stories/button.stories.tsx [hunk 21]
```diff
@@ -1,20 +1,45 @@
...
+const Template: ComponentStory<typeof IonButton> = (args: ButtonProps) => (
+  <IonButton {...args} />
 );
 
-export const ButtonDefault = Template.bind({});
+export const ButtonPrimary = Template.bind({});
+export const ButtonSecondary = Template.bind({});
+export const ButtonGhost = Template.bind({});
+export const ButtonDashed = Template.bind({});
+
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:18]: Create a case with disabled button and other with danger true

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: src/components/button/styles.ts [hunk 22]
```diff
@@ -2,22 +2,160 @@ import stitches from '../../stitches.config';
...
+) {
+  let styles: ColorStyle = {
+    background: bgColor,
+    color: fontColor,
+  };
+
+  borderColor && (styles.border = `1px solid ${borderColor}`);
+
+  return styles;
+}
```
[the remark this answers is not in the document]
Comment (Deeved) [reply src/components/button/styles.ts:34]: I think it's a good idea. so we can use it in other components.

File: package.json [hunk 23]
```diff
@@ -102,6 +102,7 @@
   },
   "dependencies": {
     "@stitches/react": "^1.2.8",
+    "@testing-library/jest-dom": "^5.16.5",
```
Comment (iurynogueira) [inline package.json:105]: Its a dev dependence

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: src/components/button/button.test.tsx [hunk 24]
```diff
@@ -1,12 +1,63 @@
...
 describe('Button', () => {
-  test('renders the Button component', () => {
+  test('should render the Button component', () => {
     const label = 'Hello world!';
     render(<Button label={label} />);
     expect(screen.queryAllByText(label)).toHaveLength(1);
   });
+
+  test('should execute event when the button is clicked', () => {
+    fireEvent.click(sut({  ...defaultButton }));
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:29]:
```suggestion
    userEvent.click(sut({  ...defaultButton }));
```

change to userEvent, fireEvent not is recommended.

File: src/components/button/button.test.tsx [hunk 25]
```diff
@@ -1,12 +1,63 @@
...
+    'secondary',
+    'ghost',
+    'dashed',
+  ];
+
+  test.each(buttonTypes)(
+    'should render button with %s style type',
+    (type) => {
+      const {className} = sut({ ...defaultButton, type: type }); 
+      expect(className).toContain(`type-${type}`);
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:52]: very well

File: src/components/button/button.tsx [hunk 26]
```diff
@@ -1,12 +1,33 @@
...
+  isDanger?: boolean;
+} & StitchesButtonProps;
 
-const Button = ({ label }: ButtonProps) => {
-  return <StitcheButton>{label}</StitcheButton>;
+const IonButton = ({
+  type = 'primary',
+  disabled = false,
+  isDanger = false,
+  onClick,
```
Comment (iurynogueira) [inline src/components/button/button.tsx:18]: I believe that we can change to handleClick to not cause confusing with the prop name.

```suggestion
  handleClick,
```

File: src/components/utils/generateColorStyle.tsx [hunk 27]
```diff
@@ -0,0 +1,30 @@
+// Tipo para objeto de estilos  da função geradora de estilos
+type ColorStyle = {
+    background: string;
+    color: string;
+    border?: string;
+  };
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:6]: check indent

File: src/components/utils/generateColorStyle.tsx [hunk 28]
```diff
@@ -0,0 +1,30 @@
+// Tipo para objeto de estilos  da função geradora de estilos
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:1]: Keep english

File: src/components/utils/generateColorStyle.tsx [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+  
+  /* 
+    Função que gera cores de Background, Border e cor de
+    fonte a fim de reduzir repetições de codigo
+  
+    Seguir o padrão de parâmetros: 
+    1- Cor do background;
+    2- Cor da fonte do botão;
+    3- (Opicional) Cor da borda;
+  */
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:16]:
```suggestion
  /* 
    Função que gera cores de Background, Border e cor de
    fonte a fim de reduzir repetições de codigo
  
    Seguir o padrão de parâmetros: 
    1- Cor do background;
    2- Cor da fonte do botão;
    3- (Opicional) Cor da borda;
  */
```

if you are going to comment out a function use the pattern.
https://tsdoc.org/

File: src/components/utils/generateColorStyle.tsx [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+  ) {
+    let styles: ColorStyle = {
+      background: bgColor,
+      color: fontColor,
+    };
+  
+    borderColor && (styles.border = `1px solid ${borderColor}`);
+  
+    return styles;
+  }
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:30]:
```suggestion
  }
  
```

missing whiteline

File: src/components/utils/generateColorStyle.tsx [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+    bgColor: string,
+    fontColor: string,
+    borderColor: string
+  ) {
+    let styles: ColorStyle = {
+      background: bgColor,
+      color: fontColor,
+    };
+  
+    borderColor && (styles.border = `1px solid ${borderColor}`);
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:27]:
```suggestion
    borderColor && (styles.border = `1px solid ${borderColor}`);
```

Read the styleguide
https://github.com/Brisanet/typescript/blob/main/docs/Style%20Guide.md#control-statements

File: src/components/utils/generateColorStyle.tsx [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+  /* 
+    Função que gera cores de Background, Border e cor de
+    fonte a fim de reduzir repetições de codigo
+  
+    Seguir o padrão de parâmetros: 
+    1- Cor do background;
+    2- Cor da fonte do botão;
+    3- (Opicional) Cor da borda;
+  */
+ export function generateColorStyle(
```
Comment (iurynogueira) [inline src/components/utils/generateColorStyle.tsx:17]:
```suggestion
 export function getBgFontBorderColors(
```

I believe that this name make more sense to the return of your function

File: src/components/button/button.tsx [hunk 33]
```diff
@@ -1,12 +1,33 @@
...
+import { ButtonStyles } from './styles';
 
-export interface ButtonProps extends InputHTMLAttributes<HTMLButtonElement> {
+type StitchesButtonProps = React.ComponentProps<typeof ButtonStyles>;
+
+export type ButtonProps = {
   label: string;
-}
+  type?: string;
+  disabled?: boolean;
```
Comment (iurynogueira) [inline src/components/button/button.tsx:9]:
```suggestion
  disabled?: boolean = false;
```

implicity this value

File: src/stories/button.stories.tsx [hunk 34]
```diff
@@ -1,20 +1,58 @@
...
+export const PrimaryDisabled = Template.bind({});
+
+export const PrimaryDanger = Template.bind({});
+
+// Button types
+ButtonPrimary.args = {
+  label: 'Primary',
+  type: 'primary',
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:27]:
```suggestion
  disabled: false,
```
bike shedding

File: src/stories/button.stories.tsx [hunk 35]
```diff
@@ -1,20 +1,58 @@
...
+};
+ButtonSecondary.args = {
+  label: 'Secondary',
+  type: 'secondary',
+  disabled: false,
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:32]: bike shedding

File: src/stories/button.stories.tsx [hunk 36]
```diff
@@ -1,20 +1,58 @@
...
+};
+ButtonDashed.args = {
+  label: 'Dashed',
+  type: 'dashed',
+  disabled: false,
+};
+
+// Button disabled
+PrimaryDisabled.args = {
```
Comment (iurynogueira) [inline src/stories/button.stories.tsx:46]:
```suggestion
PrimaryDisabledAndDanger.args = {
```

File: src/components/button/button.test.tsx [hunk 37]
```diff
@@ -1,12 +1,63 @@
...
+  });
+
+  test('should be disabled', () => {
+    expect(sut({ ...defaultButton, disabled: true })).toHaveAttribute(
+      'disabled'
+    );
+  });
 });
+
+describe('Button Props', () => {
```
Comment (iurynogueira) [inline src/components/button/button.test.tsx:40]:
```suggestion
describe('Button Types', () => {
```

Comment (iurynogueira) [reply src/components/button/button.test.tsx:40]: because you are checking only the types

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: src/components/button/button.test.tsx [hunk 38]
```diff
@@ -1,12 +1,63 @@
...
+): HTMLElement {
+  render(<Button {...props} />);
+  return screen.getByRole('button', { name: props.label });
+}
 
 describe('Button', () => {
-  test('renders the Button component', () => {
+  test('should render the Button component', () => {
     const label = 'Hello world!';
     render(<Button label={label} />);
```
Comment (Deeved) [inline src/components/button/button.test.tsx:24]:
```suggestion
    sut({ label });
```

File: src/components/utils/getBgFontBorderColors.ts [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+/**
+ * Function to prevent repetition of style blocks.
+ *
+ * @param bgColor - Background color
+ * @param fontColor - The font color
+ * @param borderColor - Optional param to add a border color
+ * @returns An Javascript object containing color properties.
+ */
```
Comment (iurynogueira) [inline src/components/utils/getBgFontBorderColors.ts:14]: 👏

File: src/components/utils/getBgFontBorderColors.ts [hunk 40]
```diff
@@ -0,0 +1,27 @@
...
+export function getBgFontBorderColors(
+  background: string,
+  font: string,
+  border: string
+) {
+  let styles: ColorStyle = {
+    background: background,
+    color: font,
+    [border]: border,
```
Comment (iurynogueira) [inline src/components/utils/getBgFontBorderColors.ts:23]:
```suggestion
    border: border,
```

**Review by iurynogueira — APPROVED** [review]

**Pull request merged**